resteasy-core/src/main/java/org/jboss/resteasy/specimpl/BuiltResponse.java

```diff
@@ -38,488 +31,254 @@
-      return getHeaderValueProcessor().toHeaderString(header);
-   }
 
-   @Override
-   public MultivaluedMap<String, String> getStringHeaders()
-   {
-      CaseInsensitiveMap<String> map = new CaseInsensitiveMap<String>();
-      for (Map.Entry<String, List<Object>> entry : metadata.entrySet())
+      }
+      catch (ProcessingException pe)
       {
-         for (Object obj : entry.getValue())
-         {
-            map.add(entry.getKey(), toHeaderString(obj));
-         }
+         throw pe;
       }
-      return map;
-   }
-
-   @Override
-   public String getHeaderString(String name)
-   {
-      List vals = metadata.get(name);
-      if (vals == null) return null;
-      StringBuilder builder = new StringBuilder();
-      boolean first = true;
-      for (Object val : vals)
+      catch (Exception ex)
       {
-         if (first) first = false;
-         else builder.append(",");
-         if (val == null) val = "";
-         val = toHeaderString(val);
-         if (val == null) val = "";
-         builder.append(val);
+         throw new ProcessingException(ex);
       }
-      return builder.toString();
    }
 
-   @Override
-   public URI getLocation()
+   protected InputStream getEntityStream()
    {
-      Object uri = metadata.getFirst(HttpHeaders.LOCATION);
-      if (uri == null) return null;
-      if (uri instanceof URI) return (URI)uri;
-      String str = null;
-      if (uri instanceof String) str = (String)uri;
-      else str = toHeaderString(uri);
-      return URI.create(str);
+      if (bufferedEntity != null) {
+         return new ByteArrayInputStream(bufferedEntity);
+      }
+      if (isClosed()) {
+         throw new ProcessingException(Messages.MESSAGES.streamIsClosed());
+      }
+      InputStream is = getInputStream();
+      return is != null ? new AbstractBuiltResponse.InputStreamWrapper<BuiltResponse>(is, this) : null;
    }
 
-   @Override
-   public Set<Link> getLinks()
+   protected void setInputStream(InputStream is)
    {
-      LinkHeaders linkHeaders = getLinkHeaders();
-      Set<Link> links = new HashSet<Link>();
-      links.addAll(linkHeaders.getLinks());
-      return links;
+      this.is = is;
    }
 
-   private LinkHeaders getLinkHeaders()
+   protected InputStream getInputStream()
    {
-      LinkHeaders linkHeaders = new LinkHeaders();
-      linkHeaders.addLinkObjects(metadata, getHeaderValueProcessor());
-      return linkHeaders;
-   }
-
-   @Override
-   public boolean hasLink(String relation)
-   {
-      return getLinkHeaders().getLinkByRelationship(relation) != null;
+      if (is == null && entity != null && entity instanceof InputStream) {
+         is = (InputStream) entity;
+      }
+      return is;
    }
 
    @Override
-   public Link getLink(String relation)
+   public void releaseConnection() throws IOException
    {
-      return getLinkHeaders().getLinkByRelationship(relation);
+      releaseConnection(false);

```

Comment:
@rsearls, I think releaseConnection() could be removed from ApacheHttpClient43Engine$ClientResponse() [not sure of the how to name that inner class.]

Reply:
No, releaseConnection() can't be removed from ApacheHttpClient43Engine.
This code performs tasks very different from BuiltResponse.releaseConnection()
A bunch of test break when this code is removed.

Reply:
Ah, sorry, I see now that they're not the same.